ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,313 @@
+#include <string.h>
+#include "Locks.hpp"
+#include "VnodeCache.hpp"
+#include "Memory.hpp"
+#include "KextLog.hpp"
+#include "../PrjFSKext/public/PrjFSCommon.h"
+
+#include "VnodeCachePrivate.hpp"
+
+#ifdef KEXT_UNIT_TESTING
+#include "VnodeCacheTestable.hpp"
+#endif
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked();
+KEXT_STATIC_INLINE uintptr_t HashVnode(vnode_t _Nonnull vnode);
+
+KEXT_STATIC bool TryGetVnodeRootFromCache(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle);
+
+KEXT_STATIC void UpdateCacheForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vfs_context_t _Nonnull context,
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    bool invalidateEntry,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle);
+
+KEXT_STATIC bool TryFindVnodeIndex_Locked(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    /* out parameters */
+    uintptr_t& vnodeIndex);
+
+KEXT_STATIC bool TryUpdateEntryToLatest_ExclusiveLocked(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    bool invalidateEntry,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle);
+
+KEXT_STATIC uint32_t s_entriesCapacity;
+KEXT_STATIC VnodeCacheEntry* s_entries;
+static RWLock s_entriesLock;
+static const uint32_t MinEntriesCapacity = 0x040000; //  4 MB (assuming 16 bytes per VnodeCacheEntry)
+static const uint32_t MaxEntriesCapacity = 0x400000; // 64 MB (assuming 16 bytes per VnodeCacheEntry)
+
+kern_return_t VnodeCache_Init()
+{
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+    
+    s_entriesLock = RWLock_Alloc();
+    if (!RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+
+    s_entriesCapacity = Clamp(desiredvnodes * 2u, MinEntriesCapacity, MaxEntriesCapacity);
```

Comment:
I'm sure I'm just missing it but where is desiredvnodes defined?

Reply:
It's declared in [vnode.h](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/sys/vnode.h#L815) and defined in [param.c](https://github.com/apple/darwin-xnu/blob/0a798f6738bc1db01281fc08ae024145e84df927/bsd/conf/param.c#L107)